tracker/peerstore/local.go

```diff
@@ -0,0 +1,228 @@
+	if !ok {
+		return nil, nil
+	}
+
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+
+	if len(g.peers) < n {
+		n = len(g.peers)
+	}
+	result := make([]*core.PeerInfo, 0, n)
+
+	// We rely on random map iteration to pick n random peers.
+	for id, p := range g.peers {
+		result = append(result, core.NewPeerInfo(id, p.ip, p.port, false /* origin */, p.complete))
+		if len(result) == n {
+			break
+		}
+	}
+	return result, nil
+}
+
+// UpdatePeer implements Store.
+func (s *LocalStore) UpdatePeer(h core.InfoHash, p *core.PeerInfo) error {
+	g := s.getOrInitLockedPeerGroup(h)
+	defer g.mu.Unlock()
+
+	expiresAt := s.clk.Now().Add(s.config.TTL)
+
+	g.peers[p.PeerID] = &peerEntry{
+		ip:        p.IP,
+		port:      p.Port,
+		complete:  p.Complete,
+		expiresAt: expiresAt,
+	}
+
+	// Allows cleanupExpiredPeerGroups to quickly determine when the last
+	// peerEntry expires.
+	g.lastExpiresAt = expiresAt
+
+	return nil
+}
+
+func (s *LocalStore) getOrInitLockedPeerGroup(h core.InfoHash) *peerGroup {
+	// We must take care to handle a race condition against
+	// cleanupExpiredPeerGroups. Consider two goroutines, A and B, where A
+	// executes getOrInitLockedPeerGroup and B executes
+	// cleanupExpiredPeerGroups:
+	//
+	// A: locks s.mu, reads g from s.peerGroups, unlocks s.mu
+	// B: locks s.mu, locks g.mu, deletes g from s.peerGroups, unlocks g.mu
+	// A: locks g.mu
+	//
+	// At this point, A is holding onto a peerGroup reference which has been
+	// deleted from the peerGroups map, and thus has no choice but to attempt to
+	// reload a new peerGroup. Since the cleanup interval is quite large, it is
+	// *extremely* unlikely this for-loop will execute more than twice.
+	for {
+		s.mu.Lock()
+		g, ok := s.peerGroups[h]
+		if !ok {
+			g = &peerGroup{
+				peers:         make(map[core.PeerID]*peerEntry),
+				lastExpiresAt: s.clk.Now().Add(s.config.TTL),
+			}
+			s.peerGroups[h] = g
+		}
+		s.mu.Unlock()
+
+		g.mu.Lock()
+		if g.deleted {
+			g.mu.Unlock()
+			continue
+		}
+		return g
+	}
+}
+
+func (s *LocalStore) cleanupTask() {
+	for {
+		select {
+		case <-s.cleanupExpiredPeerEntriesTicker.C:
+			s.cleanupExpiredPeerEntries()
+		case <-s.cleanupExpiredPeerGroupsTicker.C:
+			s.cleanupExpiredPeerGroups()
+		case <-s.stop:
+			return
+		}
+	}
+}
+
+func (s *LocalStore) cleanupExpiredPeerEntries() {
+	s.mu.RLock()
+	hashes := make([]core.InfoHash, 0, len(s.peerGroups))
+	for h := range s.peerGroups {
+		hashes = append(hashes, h)
+	}
+	s.mu.RUnlock()
+
+	for _, h := range hashes {
+		s.mu.RLock()
+		g, ok := s.peerGroups[h]
+		s.mu.RUnlock()
+		if !ok {
+			continue
+		}
+
+		var expired []core.PeerID
+		g.mu.RLock()
+		for id, p := range g.peers {
+			if s.clk.Now().After(p.expiresAt) {
+				expired = append(expired, id)
+			}
+		}
+		g.mu.RUnlock()
+
+		if len(expired) == 0 {
+			// Fast path -- no need to acquire a write lock if there are no
+			// expired entries.
+			continue
+		}
+		g.mu.Lock()
+		for _, id := range expired {
+			delete(g.peers, id)
```

Comment:
So I think there's a race condition here where UpdatePeers could update the expiresAt field between the time we release the read lock and acquire the write lock. It seems like we should recheck the expiredAt value here before we actually perform the delete.

Reply:
Indeed. Nice catch.